markymark/src/main/java/com/m2mobi/markymark/composable/ListItem.kt

```diff
@@ -0,0 +1,174 @@
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Path
+import androidx.compose.ui.graphics.drawscope.DrawScope
+import androidx.compose.ui.graphics.drawscope.Fill
+import androidx.compose.ui.graphics.drawscope.Stroke
+import androidx.compose.ui.graphics.drawscope.rotate
+import androidx.compose.ui.platform.LocalDensity
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.coerceAtLeast
+import androidx.compose.ui.unit.dp
+import com.m2mobi.markymark.composer.padding
+import com.m2mobi.markymark.model.AnnotatedStableNode
+import com.m2mobi.markymark.model.ComposableStableNode.ListItemType
+import com.m2mobi.markymark.model.ComposableStableNode.ListItemType.Ordered
+import com.m2mobi.markymark.model.ComposableStableNode.ListItemType.Task
+import com.m2mobi.markymark.model.ComposableStableNode.ListItemType.Unordered
+import com.m2mobi.markymark.theme.ListBlockStyle
+import com.m2mobi.markymark.theme.OrderedListItemStyle
+import com.m2mobi.markymark.theme.TaskListItemStyle
+import com.m2mobi.markymark.theme.UnorderedListItemStyle
+import com.m2mobi.markymark.theme.UnorderedListItemStyle.Indicator
+import com.m2mobi.markymark.theme.UnorderedListItemStyle.Indicator.Shape.Oval
+import com.m2mobi.markymark.theme.UnorderedListItemStyle.Indicator.Shape.Rectangle
+import com.m2mobi.markymark.theme.UnorderedListItemStyle.Indicator.Shape.Triangle
+import com.m2mobi.markymark.theme.UnorderedListItemStyle.Indicator.Style
+
+@Composable
+internal fun ListItem(
+    modifier: Modifier,
+    type: ListItemType,
+    children: List<AnnotatedStableNode>,
+    blockStyle: ListBlockStyle,
+    level: Int,
+) {
+    val style = when (type) {
+        is Ordered -> blockStyle.orderedStyle
+        is Unordered -> blockStyle.unorderedStyle
+        is Task -> blockStyle.taskStyle
+    }
+    Row(modifier = modifier.padding(style.padding)) {
+        val density = LocalDensity.current
+        var firstLinePos by remember { mutableStateOf(0.dp to 0.dp) }
+
+        when (type) {
+            is Ordered -> OrderedIndicator(
+                type = type,
+                style = blockStyle.orderedStyle,
+            )
+            is Unordered -> UnorderedIndicator(
+                firstLinePos = firstLinePos,
+                level = level,
+                style = blockStyle.unorderedStyle,
+            )
+            is Task -> TaskIndicator(
+                type = type,
+                style = blockStyle.taskStyle,
+            )
+        }
+
+        TextNode(
+            nodes = children,
+            style = style.textStyle,
+        ) {
+            density.run {
+                if (it.lineCount > 0) firstLinePos = it.getLineTop(0).toDp() to it.getLineBottom(0).toDp()
```

Comment:
This type of behaviour can be achieved with `SubcomposeLayout` (replacing the Row) - using the measurements of a component to affect the layout of another component:

You need the overload without the state for this case. Inside it's `measurePolicy` block you have a `SubcomposeMeasureScope` with a single `fun subcompose(slotId: Any?, content: @Composable () -> Unit): List<Measurable>`:

`onTextLayout` is called when a Text component is measured so you can get the value before the indicator composes:
```
SubcomposeLayout(
  modifier = modifier, // constraints given to measurepolicy
  measurePolicy = { constraints ->
     var lineCount = 0
     val textNodeMeasurables = subcompose("textNode") { composableHere(); lineCount = ... }
     val textNode = textNodeMeasurables.first().measure(constraints) // This triggers onTextLayout
     
     val indicatorMeasurables = subcompose("indicator") { composableHere(lineCount) }
     val indicator = indicatorMeasurables.first().measure(constraints)
 
     layout(width,height) {
        textNode.placeRelative(...)
        indicator.placeRelative(...)
     }
  }
)
```